### PR TITLE
Merge CARMA Streets Plugin

### PR DESCRIPTION
**Pull request by paulbourelly999**
# PR Details
## Description
Merge CARMA-Streets plugin into develop.  Provides additional functionality required for CARMA-Streets. Acts as message forwarding for incoming BSM, Mobility Path and Mobility Operation messages to CARMA-Streets services. Also forwards outgoing Mobility Operation messages to vehicles from CARMA Streets

## Motivation and Context
Enable vehicle communication with CARMA-Streets services

## How Has This Been Tested?
Interesting testing 

## Types of changes

- [ ] Defect fix (non-breaking change that fixes an issue)
- [ x ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [ ] I have added any new packages to the sonar-scanner.properties file
- [ x ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ x ] I have read the **CONTRIBUTING** document.
[V2XHUB Contributing Guide](https://github.com/usdot-fhwa-OPS/V2X-Hub/blob/develop/Contributing.md) 
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: src/v2i-hub/CARMAStreetsPlugin/src/CARMAStreetsPlugin.cpp [hunk 1]
```diff
@@ -0,0 +1,701 @@
...
+
+			if (produce_error == RdKafka::ERR_NO_ERROR) {
+				FILE_LOG(logDEBUG) <<"Queued message:" << message;
+				retry = false;
+			}
+			else 
+			{
+				FILE_LOG(logERROR) <<"Failed to queue message:" << message <<" with error:" << RdKafka::err2str(produce_error);
+				if (produce_error == RdKafka::ERR__QUEUE_FULL) {
+					FILE_LOG(logERROR) <<"Message queue full...retrying...";
```
Comment (paulbourelly999) [inline src/v2i-hub/CARMAStreetsPlugin/src/CARMAStreetsPlugin.cpp:470]: PLOG

File: src/v2i-hub/CARMAStreetsPlugin/src/CARMAStreetsPlugin.cpp [hunk 2]
```diff
@@ -0,0 +1,701 @@
...
+																		message.size(), 
+																		NULL, NULL, 0, 0);
+
+			if (produce_error == RdKafka::ERR_NO_ERROR) {
+				FILE_LOG(logDEBUG) <<"Queued message:" << message;
+				retry = false;
+			}
+			else 
+			{
+				FILE_LOG(logERROR) <<"Failed to queue message:" << message <<" with error:" << RdKafka::err2str(produce_error);
```
Comment (paulbourelly999) [inline src/v2i-hub/CARMAStreetsPlugin/src/CARMAStreetsPlugin.cpp:468]: PLOG

File: src/v2i-hub/CARMAStreetsPlugin/src/CARMAStreetsPlugin.cpp [hunk 3]
```diff
@@ -0,0 +1,701 @@
...
+		{
+			RdKafka::ErrorCode produce_error = kafka_producer->produce(_transmitBSMTopic, 
+																		RdKafka::Topic::PARTITION_UA,
+																		RdKafka::Producer::RK_MSG_COPY, 
+																		const_cast<char *>(message.c_str()),
+																		message.size(), 
+																		NULL, NULL, 0, 0);
+
+			if (produce_error == RdKafka::ERR_NO_ERROR) {
+				FILE_LOG(logDEBUG) <<"Queued message:" << message;
```
Comment (paulbourelly999) [inline src/v2i-hub/CARMAStreetsPlugin/src/CARMAStreetsPlugin.cpp:463]: PLOG

File: src/v2i-hub/CARMAStreetsPlugin/src/CARMAStreetsPlugin.cpp [hunk 4]
```diff
@@ -0,0 +1,701 @@
...
+		if (produce_error == RdKafka::ERR_NO_ERROR) 
+		{
+			FILE_LOG(logDEBUG) << "Queued message:" << json_message;
+		}
+		else 
+		{
+			FILE_LOG(logERROR) << "Failed to queue message:" << json_message <<" with error:" << RdKafka::err2str(produce_error);
+			if (produce_error == RdKafka::ERR__QUEUE_FULL) 
+			{
+				FILE_LOG(logERROR) << "MobilityPath producer Message queue is full.";
```
Comment (paulbourelly999) [inline src/v2i-hub/CARMAStreetsPlugin/src/CARMAStreetsPlugin.cpp:298]: PLOG

File: src/v2i-hub/CARMAStreetsPlugin/src/CARMAStreetsPlugin.cpp [hunk 5]
```diff
@@ -0,0 +1,701 @@
...
+																			RdKafka::Producer::RK_MSG_COPY, const_cast<char *>(json_message.c_str()),
+																			json_message.size(), NULL, NULL, 0, 0 );
+
+		if (produce_error == RdKafka::ERR_NO_ERROR) 
+		{
+			FILE_LOG(logDEBUG) << "Queued message:" << json_message;
+		}
+		else 
+		{
+			FILE_LOG(logERROR) << "Failed to queue message:" << json_message <<" with error:" << RdKafka::err2str(produce_error);
```
Comment (paulbourelly999) [inline src/v2i-hub/CARMAStreetsPlugin/src/CARMAStreetsPlugin.cpp:295]: PLOG

File: src/v2i-hub/CARMAStreetsPlugin/src/CARMAStreetsPlugin.cpp [hunk 6]
```diff
@@ -0,0 +1,701 @@
...
+		const std::string json_message 			= Json::writeString(builder, mobilityPathJsonRoot);
+		PLOG(logDEBUG) <<"MobilityPath Json message:" << json_message;
+		RdKafka::ErrorCode produce_error 		= kafka_producer->produce(	_transmitMobilityPathTopic, 
+																			RdKafka::Topic::PARTITION_UA,
+																			RdKafka::Producer::RK_MSG_COPY, const_cast<char *>(json_message.c_str()),
+																			json_message.size(), NULL, NULL, 0, 0 );
+
+		if (produce_error == RdKafka::ERR_NO_ERROR) 
+		{
+			FILE_LOG(logDEBUG) << "Queued message:" << json_message;
```
Comment (paulbourelly999) [inline src/v2i-hub/CARMAStreetsPlugin/src/CARMAStreetsPlugin.cpp:291]: PLOG

File: src/v2i-hub/CARMAStreetsPlugin/src/CARMAStreetsPlugin.cpp [hunk 7]
```diff
@@ -0,0 +1,701 @@
...
+					else {
+						FILE_LOG(logERROR) <<"Unhandled error in queue_kafka_message:" << RdKafka::err2str(produce_error);
+						retry = false;
+					}
+				}	
+			}
+		}
+	}
+	catch (TmxException &ex) {
+		FILE_LOG(logERROR) << "Failed to decode message : " << ex.what();
```
Comment (paulbourelly999) [inline src/v2i-hub/CARMAStreetsPlugin/src/CARMAStreetsPlugin.cpp:204]: PLOH

File: src/v2i-hub/CARMAStreetsPlugin/src/CARMAStreetsPlugin.cpp [hunk 8]
```diff
@@ -0,0 +1,701 @@
...
+				else 
+				{
+					FILE_LOG(logERROR) <<"Failed to queue message:" << message <<" with error:" << RdKafka::err2str(produce_error);
+					if (produce_error == RdKafka::ERR__QUEUE_FULL) {
+						FILE_LOG(logERROR) <<"Message queue full...retrying...";
+						kafka_producer->poll(500);  /* ms */
+						retry = true;
+					}
+					else {
+						FILE_LOG(logERROR) <<"Unhandled error in queue_kafka_message:" << RdKafka::err2str(produce_error);
```
Comment (paulbourelly999) [inline src/v2i-hub/CARMAStreetsPlugin/src/CARMAStreetsPlugin.cpp:196]: PLOG

File: src/v2i-hub/CARMAStreetsPlugin/src/CARMAStreetsPlugin.cpp [hunk 9]
```diff
@@ -0,0 +1,701 @@
...
+
+				if (produce_error == RdKafka::ERR_NO_ERROR) {
+					FILE_LOG(logDEBUG) <<"Queued message:" << message;
+					retry = false;
+				}
+				else 
+				{
+					FILE_LOG(logERROR) <<"Failed to queue message:" << message <<" with error:" << RdKafka::err2str(produce_error);
+					if (produce_error == RdKafka::ERR__QUEUE_FULL) {
+						FILE_LOG(logERROR) <<"Message queue full...retrying...";
```
Comment (paulbourelly999) [inline src/v2i-hub/CARMAStreetsPlugin/src/CARMAStreetsPlugin.cpp:191]: PLOG

File: src/v2i-hub/CARMAStreetsPlugin/src/CARMAStreetsPlugin.cpp [hunk 10]
```diff
@@ -0,0 +1,701 @@
...
+			{
+				RdKafka::ErrorCode produce_error = kafka_producer->produce(_transmitMobilityOperationTopic, 
+																	RdKafka::Topic::PARTITION_UA,
+																	RdKafka::Producer::RK_MSG_COPY, 
+																	const_cast<char *>(message.c_str()),
+																	message.size(), 
+																	NULL, NULL, 0, 0);
+
+				if (produce_error == RdKafka::ERR_NO_ERROR) {
+					FILE_LOG(logDEBUG) <<"Queued message:" << message;
```
Comment (paulbourelly999) [inline src/v2i-hub/CARMAStreetsPlugin/src/CARMAStreetsPlugin.cpp:184]: PLOG

File: src/v2i-hub/CARMAStreetsPlugin/src/CARMAStreetsPlugin.cpp [hunk 11]
```diff
@@ -0,0 +1,701 @@
...
+
+void CARMAStreetsPlugin::HandleMobilityOperationMessage(tsm3Message &msg, routeable_message &routeableMsg ) {
+	try 
+	{
+		auto mobilityOperation = msg.get_j2735_data();
+		bool retry = true;
+		FILE_LOG(logERROR) << "Body OperationParams : " << mobilityOperation->body.operationParams.buf;
+		FILE_LOG(logERROR) << "Body Strategy : " << mobilityOperation->body.strategy.buf;
+
+  		FILE_LOG(logERROR) <<"Queueing kafka message:topic:" << _transmitMobilityOperationTopic << " "
```
Comment (paulbourelly999) [inline src/v2i-hub/CARMAStreetsPlugin/src/CARMAStreetsPlugin.cpp:129]: PLOG and adjust log level

File: src/v2i-hub/CARMAStreetsPlugin/src/CARMAStreetsPlugin.cpp [hunk 12]
```diff
@@ -0,0 +1,701 @@
...
+	UpdateConfigSettings();
+}
+
+void CARMAStreetsPlugin::HandleMobilityOperationMessage(tsm3Message &msg, routeable_message &routeableMsg ) {
+	try 
+	{
+		auto mobilityOperation = msg.get_j2735_data();
+		bool retry = true;
+		FILE_LOG(logERROR) << "Body OperationParams : " << mobilityOperation->body.operationParams.buf;
+		FILE_LOG(logERROR) << "Body Strategy : " << mobilityOperation->body.strategy.buf;
```
Comment (paulbourelly999) [inline src/v2i-hub/CARMAStreetsPlugin/src/CARMAStreetsPlugin.cpp:127]: PLOG and adjust log level

File: src/v2i-hub/CARMAStreetsPlugin/src/CARMAStreetsPlugin.cpp [hunk 13]
```diff
@@ -0,0 +1,701 @@
...
+	PluginClient::OnConfigChanged(key, value);
+	UpdateConfigSettings();
+}
+
+void CARMAStreetsPlugin::HandleMobilityOperationMessage(tsm3Message &msg, routeable_message &routeableMsg ) {
+	try 
+	{
+		auto mobilityOperation = msg.get_j2735_data();
+		bool retry = true;
+		FILE_LOG(logERROR) << "Body OperationParams : " << mobilityOperation->body.operationParams.buf;
```
Comment (paulbourelly999) [inline src/v2i-hub/CARMAStreetsPlugin/src/CARMAStreetsPlugin.cpp:126]: PLOG and adjust log level

**Review by paulbourelly999 — COMMENTED** [review]
I will create a PR to address the changes I mentioned here. Also we need to confirm that we include this new plugin in the sonar scanner analysis.

**Review by dan-du-car — COMMENTED** [review]
Please resolve the conflict.

**Review by dan-du-car — APPROVED** [review]

**Pull request merged**